File: backend/app/routes/ai/mentor.py
import json
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload
from loguru import logger

from app.database import get_db
from app.services.auth_service import get_current_user, verify_workspace_access
from app.db_models import User, MentorSession
from app.models import (
    StartMentorSessionRequest,
    MentorRespondRequest,
    MentorSessionResponse,
)
from app.services import mentor_service

router = APIRouter()


def _session_to_response(s: MentorSession) -> MentorSessionResponse:
    return MentorSessionResponse(
        id=s.id,
        workspace_id=s.workspace_id,
        topic=s.topic,
        source_ids=s.source_ids or "[]",
        messages=s.messages or "[]",
        status=s.status,
        summary=s.summary,
        gap_report=s.gap_report,
        correct_count=s.correct_count or 0,
        total_questions=s.total_questions or 0,
        created_at=s.created_at.isoformat() if s.created_at else "",
        updated_at=s.updated_at.isoformat() if s.updated_at else "",
    )


@router.post("/start", status_code=201)
async def start_mentor_session(
    req: StartMentorSessionRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await verify_workspace_access(db, req.workspace_id, user.id)
    try:
        session, first_question = await mentor_service.start_session(
            db, req.workspace_id, user, req.topic, req.source_ids, req.context,
        )
        return {
            "session": _session_to_response(session).model_dump(),
            "first_question": first_question,
        }
    except Exception as e:
        logger.exception("Failed to start mentor session")
        raise HTTPException(status_code=500, detail={"error": "mentor_start_failed", "message": str(e)})


@router.post("/respond")
async def respond_mentor(
    req: MentorRespondRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    try:
        result = await mentor_service.respond(db, user, req.session_id, req.answer)
        return result
    except ValueError as e:
        raise HTTPException(status_code=404, detail={"error": "not_found", "message": str(e)})
    except Exception as e:
        logger.exception("Failed to process mentor response")
        raise HTTPException(status_code=500, detail={"error": "mentor_respond_failed", "message": str(e)})


@router.post("/{session_id}/end")
async def end_mentor_session(
    session_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    try:
        result = await mentor_service.end_session(db, user, session_id)
        return result
    except ValueError as e:
        raise HTTPException(status_code=404, detail={"error": "not_found", "message": str(e)})
    except Exception as e:
        logger.exception("Failed to end mentor session")
        raise HTTPException(status_code=500, detail={"error": "mentor_end_failed", "message": str(e)})


@router.get("/sessions")
async def list_mentor_sessions(
    workspace_id: str = Query(...),
    status: str | None = Query(None),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await verify_workspace_access(db, workspace_id, user.id)
    query = select(MentorSession).where(
        MentorSession.workspace_id == workspace_id,
        MentorSession.user_id == user.id,
    )
    if status:
        query = query.where(MentorSession.status == status)
    query = query.order_by(MentorSession.created_at.desc())
    result = await db.execute(query)
    sessions = result.scalars().all()
    return [_session_to_response(s).model_dump() for s in sessions]


@router.get("/{session_id}")
async def get_mentor_session(
    session_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(MentorSession).where(
            MentorSession.id == session_id,
            MentorSession.user_id == user.id,
        )
    )
    session = result.scalar_one_or_none()
    if not session:
        raise HTTPException(status_code=404, detail={"error": "not_found", "message": "Mentor session not found"})
    return _session_to_response(session).model_dump()


@router.delete("/{session_id}")
async def delete_mentor_session(
    session_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(MentorSession).where(
            MentorSession.id == session_id,
            MentorSession.user_id == user.id,
        )
    )
    session = result.scalar_one_or_none()
    if not session:
        raise HTTPException(status_code=404, detail={"error": "not_found", "message": "Mentor session not found"})
    await db.delete(session)
    await db.commit()
    return {"deleted": True}

File: frontend/src/api/mentor.ts
import { apiFetch } from './client';

export interface MentorMessage {
  role: 'ai' | 'user';
  content: string;
  evaluation?: string | null;
  explanation?: string;
  correct_answer?: string | null;
  assessment?: string | null;
}

export interface MentorSessionItem {
  id: string;
  workspace_id: string;
  topic: string;
  source_ids: string;
  messages: string;
  status: string;
  summary: string | null;
  gap_report: string | null;
  correct_count: number;
  total_questions: number;
  created_at: string;
  updated_at: string;
}

export interface StartSessionResult {
  session: MentorSessionItem;
  first_question: string;
}

export interface RespondResult {
  session_id: string;
  evaluation: string | null;
  explanation: string;
  follow_up_question: string;
  correct_answer: string | null;
  assessment: string | null;
  total_questions: number;
  correct_count: number;
  session_complete: boolean;
}

export interface EndSessionResult {
  session_id: string;
  summary: string;
  gap_report: Array<{ concept: string; explanation: string; suggested_review: string }>;
  topics_covered: string[];
  correct_count: number;
  total_questions: number;
  accuracy_percentage: number;
  strengths: string[];
  weaknesses: string[];
  recommended_focus: string;
}

export async function startMentorSession(
  workspaceId: string,
  topic: string,
  sourceIds: string[] = [],
  context: string = '',
): Promise<StartSessionResult> {
  return apiFetch<StartSessionResult>('/ai/mentor/start', {
    method: 'POST',
    body: JSON.stringify({
      workspace_id: workspaceId,
      topic,
      source_ids: sourceIds,
      context,
    }),
  });
}

export async function respondMentor(
  sessionId: string,
  answer: string,
): Promise<RespondResult> {
  return apiFetch<RespondResult>('/ai/mentor/respond', {
    method: 'POST',
    body: JSON.stringify({ session_id: sessionId, answer }),
  });
}

export async function endMentorSession(
  sessionId: string,
): Promise<EndSessionResult> {
  return apiFetch<EndSessionResult>(`/ai/mentor/${sessionId}/end`, {
    method: 'POST',
  });
}

export async function fetchMentorSessions(
  workspaceId: string,
  status?: string,
): Promise<MentorSessionItem[]> {
  const params = new URLSearchParams({ workspace_id: workspaceId });
  if (status) params.set('status', status);
  return apiFetch<MentorSessionItem[]>(`/ai/mentor/sessions?${params}`);
}

export async function fetchMentorSession(
  sessionId: string,
): Promise<MentorSessionItem> {
  return apiFetch<MentorSessionItem>(`/ai/mentor/${sessionId}`);
}

export async function deleteMentorSession(
  sessionId: string,
): Promise<{ deleted: boolean }> {
  return apiFetch<{ deleted: boolean }>(`/ai/mentor/${sessionId}`, {
    method: 'DELETE',
  });
}
